Math/NumberTheory/Primes/Factorisation/QuadraticSieve.hs

```diff
@@ -5,182 +5,277 @@
 {-# LANGUAGE ScopedTypeVariables #-}
 
 module Math.NumberTheory.Primes.Factorisation.QuadraticSieve
-  ( quadraticSieve
+  ( QuadraticSieveConfig(..)
+  , quadraticSieve
+  , quadraticSieveManual
   , findSquares
   ) where
 
 #if __GLASGOW_HASKELL__ < 803
 import Data.Semigroup
 #endif
+import qualified Data.Set as SS
 import qualified Data.List as L
 import qualified Data.IntMap as I
+import qualified Data.IntSet as S
 import qualified Data.Vector as V
 import qualified Data.Vector.Sized as SV
 import qualified Data.Vector.Mutable as MV
 import qualified Data.Vector.Unboxed as U
 import qualified Data.Vector.Unboxed.Sized as SU
-import qualified Math.NumberTheory.Primes.IntSet as PS
+import qualified Data.Mod as M
+import qualified Data.Mod.Word as MW
 import Math.NumberTheory.Roots
 import Math.NumberTheory.Primes
+import Math.NumberTheory.Logarithms
 import Math.NumberTheory.Moduli.Sqrt
 import Math.NumberTheory.Utils.FromIntegral
 import Math.NumberTheory.Primes.Factorisation.LinearAlgebra
+import Math.NumberTheory.Primes.Factorisation.TrialDivision
 import Control.Monad
 import Control.Monad.ST
 import GHC.TypeNats
 import Data.Proxy
 import Data.Kind
+import Data.Foldable
 import Data.Maybe
 import Data.Bit
 import Data.Bifunctor
+import qualified Debug.Trace
 
--- This datatype stores factorisations of numbers. Its @sign@ describes the
--- sign of the number (True = Negative and False = Non-negative) and @primeSet@
--- stores the prime numbers in its factorisation.
-data SignedPrimeIntSet = SignedPrimeIntSet
-  { sign :: !Bool
-  , primeSet :: !PS.PrimeIntSet
-  } deriving (Show, Eq, Ord)
+trace :: String -> a -> a
+trace = if debug then Debug.Trace.trace else const id
 
-insert :: Prime Int -> SignedPrimeIntSet -> SignedPrimeIntSet
-insert prime (SignedPrimeIntSet s ps) = SignedPrimeIntSet s (prime `PS.insert` ps)
+debug :: Bool
+debug = False
 
--- | Given an odd positive composite Integer @n@ and Int parameters @b@ and @t@,
--- the Quadratic Sieve attempts to output @factor@, a factor of @n@. The
+data QuadraticSieveConfig = QuadraticSieveConfig
+  { qscFactorBase :: Int
+  , qscSievingInterval :: Int
+  , qscNumberOfBlocks :: Int
+  , qscSmoothThreshold :: Int
+  }
+
+autoConfig :: Integer -> QuadraticSieveConfig
+autoConfig n = QuadraticSieveConfig t m k h
+  where
+    -- + 4 for large prime variation
+    h = intLog2 t + 1
+    k = max 1 (l `div` 10)
+    m = 3 * t `div` 2
+    t
+      | l < 4    = integerToInt n `div` 2
+      | l < 8    = integerToInt $ integerSquareRoot n
+      | otherwise = (50 - l) * floor (sqrt . exp . sqrt $ log (fromInteger n) * log (log (fromInteger n)) :: Double)
```

Comment:
Again, let's avoid converting huge integers to `Double`. Approximate this calculation using [`integerLog2`](http://hackage.haskell.org/package/integer-logarithms-1.0.3/docs/Math-NumberTheory-Logarithms.html#v:integerLog2).